index.js

```diff
@@ -119,16 +94,79 @@
   //  $$type :: String
   var $$type = '@@type';
 
-  //  type :: Any -> String
+  //  RPARSE :: RegExp
+  var RPARSE = /^(?:([^/\n\r]+)\/)?([^]+?)(?:@(\d+))?$/;
+
+  //       TypeIdentifier :: (a, b, c) -> TypeIdentifier a b c
+  function TypeIdentifier(namespace, name, version) {
+    return {
+      namespace: namespace,
+      name: name,
+      version: version
+    };
+  }
+
+  //       parseTypeIdentifier :: String
+  //                           -> TypeIdentifier String? String Number?
+  function parseTypeIdentifier(s) {
+    if (!s.length) return TypeIdentifier(null, s, null);
+    var parsed = RPARSE.exec(s);
+    return TypeIdentifier(
+      parsed[1] || null,
+      parsed[2],
+      parsed[3] ? parseInt(parsed[3], 10) : null
+    );
+  }
+
+  //       parseNativeType :: Any -> TypeIdentifier Null String Null
+  function parseNativeType(x) {
+    return TypeIdentifier(
+      null,
+      Object.prototype.toString.call(x).slice('[object '.length, -']'.length),
+      null
+    );
+  }
+
+  //. ### Usage
+  //.
+  //. #### With native types
+  //.
+  //. ```javascript
+  //. > type(null);
+  //. {namespace: null, name: 'Null', version: null}
+  //. > type(true);
+  //. {namespace: null, name: 'Boolean', version: null}
+  //. > type([1, 2, 3]);
+  //. {namespace: null, name: 'Array', version: null}
+  //. ```
+  //.
+  //. #### With custom types
+  //.
+  //. ```javascript
+  //. > var IdentityTypeRep = {'@@type': 'my-package/Identity'};
+  //. > function Identity(x) { this.value = x; }
+  //. > Identity.prototype.constructor = IdentityTypeRep;
```

Comment:
Doing `function Identity(x) { return {constructor: ...}; }` exceeded the maximum line length, so I went for the prototype approach, spreading the function over multiple lines caused an "unexpected end of input".